core/ArchiveProcessor/RecordBuilder.php

```diff
@@ -0,0 +1,241 @@
+            $columnToSortByBeforeTruncation = $record->getColumnToSortByBeforeTruncation() ?? $this->columnToSortByBeforeTruncation;
+
+            $archiveProcessor->aggregateDataTableRecords(
+                $record->getName(),
+                $maxRowsInTable,
+                $maxRowsInSubtable,
+                $columnToSortByBeforeTruncation,
+                $this->columnAggregationOps
+            );
+        }
+
+        if (!empty($numericRecords)) {
+            $numericMetrics = array_map(function (Record $r) { return $r->getName(); }, $numericRecords);
+            if (!empty($requestedReports)) {
+                $numericMetrics = array_filter($numericMetrics, function ($name) use ($requestedReports, $foundRequestedReports) {
+                    return in_array($name, $requestedReports) && !in_array($name, $foundRequestedReports);
+                });
+            }
+            $archiveProcessor->aggregateNumericMetrics($numericMetrics, $this->columnAggregationOps);
+        }
+    }
+
+    /**
+     * Returns metadata for records primarily used when aggregating over non-day periods. Every numeric/blob
+     * record your RecordBuilder creates should have an associated piece of record metadata.
+     *
+     * @return Record[]
+     */
+    public abstract function getRecordMetadata(ArchiveProcessor $archiveProcessor): array;
+
+    /**
+     * Derived classes should define this method to aggregate log data for a single day and return the records
+     * to store indexed by record names.
+     *
+     * @return (DataTable|int|float|string)[] Record values indexed by their record name, eg, `['MyPlugin_MyRecord' => new DataTable()]`
+     */
+    protected abstract function aggregate(ArchiveProcessor $archiveProcessor): array;
+
+    private function insertRecord(ArchiveProcessor $archiveProcessor, $recordName, DataTable\DataTableInterface $record,
+                                  ?int $maxRowsInTable, ?int $maxRowsInSubtable, ?string $columnToSortByBeforeTruncation): void
+    {
+        $serialized = $record->getSerialized(
+            $maxRowsInTable ?: $this->maxRowsInTable,
+            $maxRowsInSubtable ?: $this->maxRowsInSubtable,
+            $columnToSortByBeforeTruncation ?: $this->columnToSortByBeforeTruncation
+        );
+        $archiveProcessor->insertBlobRecord($recordName, $serialized);
+        unset($serialized);
+    }
+
+    public function getMaxRowsInTable(): ?int
+    {
+        return $this->maxRowsInTable;
+    }
+
+    public function getMaxRowsInSubtable(): ?int
+    {
+        return $this->maxRowsInSubtable;
+    }
+
+    public function getColumnToSortByBeforeTruncation(): ?string
+    {
+        return $this->columnToSortByBeforeTruncation;
+    }
+
+    public function getPluginName(): ?string
+    {
+        $className = get_class($this);
+        $parts = explode('\\', $className);
+        $parts = array_filter($parts);
+        $plugin = $parts[2];
+        return $plugin;
```

Comment:
Is it possible for a valid class to produce a part that would evaluate to a falsey value and would needed to be filtered out?

What is the desired behaviour here, to return the last part of the class name? Or some middle part from a certain naming convention?

Reply:
Should we just have one method for this and `getQueryOriginHint`?

Reply:
> Is it possible for a valid class to produce a part that would evaluate to a falsey value and would needed to be filtered out?

Not unless a developer was doing something they definitely aren't supposed to.

> What is the desired behaviour here, to return the last part of the class name? Or some middle part from a certain naming convention?

All Matomo classes in plugins start with `Piwik\Plugins\PluginName\...`. This returns the 3rd part of that sequence, ie the name of the plugin that defines the class.

> Should we just have one method for this and getQueryOriginHint?

It should be possible for getQueryOriginHint to be overridden if desired.

> I think we should not rely on the classes structure that strictly to expect the 3rd element to be the class name without the namespaces.

I think you misunderstand what this method does?

Reply:
> All Matomo classes in plugins start with Piwik\Plugins\PluginName\.... This returns the 3rd part of that sequence, ie the name of the plugin that defines the class.

Right. So is it conceivable that a 3rd party plugin would be in `MyCompanyName\MyOrgName\Matomo\Plugins\PluginName\...` namespace?

Even if people use the console command to start a plugin, can they then change the namespaces? Or is the whole mechanism dependent on the plugin class name structure?

If that was the case, we should probably provide a method that returns a plugin name as a single line of code rather than repeating this explode mechanism in several place, as suggested in https://github.com/matomo-org/matomo/pull/20394#discussion_r1200502867

Reply:
> > Is it possible for a valid class to produce a part that would evaluate to a falsey value and would needed to be filtered out?

> Not unless a developer was doing something they definitely aren't supposed to.

Ok, so we can drop the `array_filter` part?

----

> > Should we just have one method for this and getQueryOriginHint?

> It should be possible for getQueryOriginHint to be overridden if desired.

Yep, all good, I realised that later on when looking at the specific implementations.

Reply:
> Ok, so we can drop the array_filter part?

Sure.